editoast/src/views/infra/mod.rs

```diff
@@ -569,9 +571,15 @@ async fn get_speed_limit_tags(
     })
     .await?;
     let speed_limits_tags = infra.get_speed_limit_tags(conn).await?;
-    Ok(Json(
-        speed_limits_tags.into_iter().map(|el| (el.tag)).collect(),
-    ))
+    let set_speed_limits_tags: HashSet<String> =
+        speed_limits_tags.into_iter().map(|el| (el.tag)).collect();
+    let speed_limit_tags_ids = SpeedLimitTagIds::load().0;
+    let set_speed_limit_tags_ids: HashSet<String> = speed_limit_tags_ids.into_iter().collect();
+    let all_speed_limit_tags: Vec<String> = set_speed_limit_tags_ids
+        .union(&set_speed_limits_tags)
+        .cloned()
+        .collect();
+    Ok(Json(all_speed_limit_tags))
```

Comment:
1. We can change the endpoint signature to return a `Result<Json<HashSet<String>>>`. So we don't have to transform the `HashSet` into a `Vec`.
2. Instead of building two HashSet and calling `HashSet::union`, we can just build an iterator over both list and then collect as HashSet.
3. We don't need to call `SpeedLimitTagIds::load()`. We can retrieve this list from the state `State(builtin_tags): State<SpeedLimitTagIds>`

```suggestion
    let infra_tags = infra.get_speed_limit_tags(conn).await?;
    let union_tags = infra_tags
        .into_iter()
        .map(|el| (el.tag))
        .chain(builtin_tags.0)
        .collect();
    Ok(Json(union_tags))
```

Reply:
About 3., we can't directly do that. Either extract the tags from the state directly `State(AppState { speed_limit_tags_ids, .. }): State<AppState>` or implement `FromRef<AppState> for Arc<SpeedLimitTagIds>` to have the shorthand and then you can indeed `State(builtin_tags): State<Arc<SpeedLimitTagIds>>` (note the `Arc`). Whichever you prefer is fine by me :)

Reply:
We already impl FromRef for the DB pool in views/mod.rs, you can do something similar